Disable Maven2 legacy support for `maven-publish` plugin

bigdaz opened the pull request:
When publishing with `MavenDeployer` we set the `maven.metadata.legacy` flag to ensure compatibility with Maven2. Issue #2882 is about changing this default when using `MavenDeployer`.

This PR changes the default value for the `maven-publish` plugin, which provides less strong compatibility guarantees due to it's incubating status. (Granted, `maven-publish` has been around for a long time and we endeavour to maintain compatibility: we don't think this change will cause any breakages.)

bigdaz commented:
@blindpirate I'm asking you for review since you have context from the original issue.

blindpirate commented on subprojects/maven/src/main/java/org/gradle/api/publication/maven/internal/deployer/BaseMavenDeployer.java:
```diff
@@ -51,6 +51,7 @@ public BaseMavenDeployer(PomFilterContainer pomFilterContainer, ArtifactPomConta
     protected MavenPublishAction createPublishAction(File pomFile, File metadataFile, LocalMavenRepositoryLocator mavenRepositoryLocator) {
         MavenWagonDeployAction deployAction = new MavenWagonDeployAction(pomFile, metadataFile, getJars());
         deployAction.setLocalMavenRepositoryLocation(mavenRepositoryLocator.getLocalMavenRepository());
+        deployAction.produceLegacyMavenMetadata();
```
Why should they be moved here?

bigdaz replied:
I don't understand your question.

blindpirate replied:
I mean, looks like you were moving `session.getConfigProperties().put("maven.metadata.legacy", "true");` from `AbstractMavenPublishAction` constructor to here, why is it?

bigdaz replied:
Because `AbstractMavenPublishAction` is used for all Maven publishing, via the `Upload` task _and_ via the `maven-publish` plugin. We don't want to change the default for `Upload`, but we want to disable legacy publishing via the `@Incubating` `maven-publish` plugin. So I moved this down to a common type used by `Upload`, but not `maven-publish`.

blindpirate commented on subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/maven/DefaultSnapshotMavenMetaData.groovy:
```diff
@@ -0,0 +1,65 @@
+/**
+ * http://maven.apache.org/ref/3.0.1/maven-repository-metadata/repository-metadata.html
+ */
+class DefaultSnapshotMavenMetaData implements SnapshotMavenMetaData {
```
Why not let `DefaultSnapshotMavenMetaData` extend `DefaultRootMavenMetaData`? Looks like you changed `DefaultRootMavenMetaData.version` to `DefaultRootMavenMetaData.releaseVersion` to emphasize "this is only for release metadata", but I think it's not a big deal. `DefaultRootMavenMetaData` has all that `DefaultSnapshotMavenMetaData` needs.

bigdaz replied:
The 'root' and 'snapshot' files have different content, even if they look similar.
- There is one root file for all versions. It has `groupId`, `artifactId`, list of `versions`, and latest `releaseVersion`
- There is one artifact/snapshot per version. It has `groupId`, `artifactId`, `version`, list of `snapshotVersion` and latest `snapshotVersion`

So these are similar, but I think it's clearer to keep the interfaces separate so that people understand that these are different files with different content.

blindpirate replied:
Thanks for clarifying.

blindpirate commented:
What's more, I guess this should be mentioned in release note. In case some unexpected issues occur, it would help them to figure out what happened (although we think this is transparent to users)

bigdaz commented:
@blindpirate I updated the release notes and added test coverage for sequential publishing of snapshots. Moving from `upload` to `maven-publish` (legacy->non-legacy) works as expected.

Unfortunately, when publishing multiple times with `maven-publish`, we fail to retain old `<snapshotVersions>`, which means that this does address #2882. Only the _current_ version is included in the `snapshotVersions` list.

I think it's worth merging this PR as-is, since the `maven-publish` plugin is `@incubating` and we don't want to preserve legacy support. But I'm not going to have time to address the fact that we don't include previous snapshot versions in the list.

blindpirate commented:
Maybe there's something I've misunderstood, but according to [this comment](https://github.com/gradle/gradle/issues/2882#issuecomment-328079778), all `<snapshotVersions>` should be retained in `<snapshotVersions>` list after publishing twice, right?

bigdaz commented:
> Maybe there's something I've misunderstood, but according to this comment, all <snapshotVersions> should be retained in <snapshotVersions> list after publishing twice, right?

Correct. See https://github.com/gradle/gradle/pull/3609#issuecomment-347946681 and the comment I added to https://github.com/gradle/gradle/pull/3609/files#diff-f35c961c27c40ff43c265108bb81483f.

Also, my previous comment:
> I think it's worth merging this PR as-is, since the `maven-publish` plugin is `@incubating` and we don't want to preserve legacy support. But I'm not going to have time to address the fact that we don't include previous snapshot versions in the list.
